core/platform/search/elastic_search_services.py

```diff
@@ -0,0 +1,159 @@
+# limitations under the License.
+
+"""Provides platform search services implemented using the elastic search python
+API.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import json
+
+import elasticsearch
+import feconf
+import python_utils
+
+MAXIMUM_NUMBER_OF_PAGES = 10000
+ES = elasticsearch.Elasticsearch()
+
+
+def add_documents_to_index(documents, index_name):
+    """Adds a document to an index.
+
+    Args:
+        documents: list(dict). Each document should be a dictionary.
+            Every key in the document is a field name, and the corresponding
+            value will be the field's value.
+            There MUST be a key named 'id', its value will be used as the
+            document's id.
+        index_name: str. The name of the index to insert the document into.
+
+    Raises:
+        Exception. A document cannot be added to the index.
+    """
+    assert isinstance(index_name, python_utils.BASESTRING)
+
+    for document in documents:
+        assert 'id' in document
+    for document in documents:
+        response = ES.index(
+            index_name, document, id=document['id'])
+        if response is None or response['_shards']['failed'] > 0:
+            raise Exception(
+                'Failed to add document to index.')
+
+
+def delete_documents_from_index(doc_ids, index_name):
+    """Deletes documents from an index.
+
+    Args:
+        doc_ids: list(str). A list of document ids of documents to be deleted
+            from the index.
+        index_name: str. The name of the index to delete the document from.
+
+    Raises:
+        Exception. Document id does not exist.
+    """
+    assert isinstance(index_name, python_utils.BASESTRING)
+    for doc_id in doc_ids:
+        assert isinstance(doc_id, python_utils.BASESTRING)
+
+    for doc_id in doc_ids:
+        if ES.exists(index_name, doc_id):
+            ES.delete(index_name, doc_id)
+        else:
+            raise Exception(
+                'Document id does not exist: %s' % doc_id)
+
+
+def clear_index(index_name):
+    """Clears an index on the elastic search instance.
+
+    Args:
+        index_name: str. The name of the index to clear.
+    """
+    assert isinstance(index_name, python_utils.BASESTRING)
+    # More details on clearing an index can be found here:
+    # https://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.Elasticsearch.delete_by_query
+    ES.delete_by_query(
+        index_name,
+        {
+            'query':
+                {
+                    'match_all': {}
+                }
+        })
+
+
+def search(
+        query_string, index_name, offset=None,
+        limit=feconf.SEARCH_RESULTS_PAGE_SIZE, ids_only=False):
+    """Searches for documents matching the given query in the given index.
+    NOTE: We cannot search through more than 10,000 documents by paginating
```

Comment:
Here, 10000 refers to the total number of docs, but in the args it's used as the size of a page fetched in response to a query. Which is it?

Reply:
Hm, I might have confused the definitions. The documentation uses paging and hits semi interchangeably, clarified. A hit is a single result, so in this case, looking at the definition, a single page is also the same thing(e.g an exploration). I'll just change the wording to pages. 
https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-request-from-size.html